RecoTracker/FinalTrackSelectors/plugins/TfGraphDefProducer.cc

```diff
@@ -0,0 +1,63 @@
+// -*- C++ -*-
+//
+// Package:    RecoTracker/FinalTrackSelectors
+// Class:      TFGraphDefProducer
+//
+/**\class TFGraphDefProducer
+ Description: Produces TfGraphRecord into the event containing a tensorflow GraphDef object that can be used for running inference on a pretrained network
+*/
+//
+// Original Author:  Joona Havukainen
+//         Created:  Fri, 24 Jul 2020 08:04:00 GMT
+//
+//
+
+// system include files
+#include <memory>
+
+// user include files
+#include "FWCore/Framework/interface/ModuleFactory.h"
+#include "FWCore/Framework/interface/ESProducer.h"
+
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "TrackingTools/Records/interface/TfGraphRecord.h"
+#include "RecoTracker/FinalTrackSelectors/interface/TfGraphDefWrapper.h"
+
+// class declaration
+
+class TfGraphDefProducer : public edm::ESProducer {
+public:
+  TfGraphDefProducer(const edm::ParameterSet&);
+  using ReturnType = std::unique_ptr<TfGraphDefWrapper>;
+
+  ReturnType produce(const TfGraphRecord&);
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  const std::string filename_;
+  // ----------member data ---------------------------
+};
+
+TfGraphDefProducer::TfGraphDefProducer(const edm::ParameterSet& iConfig)
+    : filename_(iConfig.getParameter<edm::FileInPath>("FileName").fullPath()) {
+  auto componentName = iConfig.getParameter<std::string>("ComponentName");
+  setWhatProduced(this, componentName);
+}
+
+// ------------ method called to produce the data  ------------
+std::unique_ptr<TfGraphDefWrapper> TfGraphDefProducer::produce(const TfGraphRecord& iRecord) {
+  return std::make_unique<TfGraphDefWrapper>(tensorflow::createSession(tensorflow::loadGraphDef(filename_), 1));
```

Comment:
Does this actually leak the `GraphDef`?

Reply:
looking into it

Reply:
Thanks for pointing it out! Per the recent CMSSW ML documentation, we need to indeed delete the graphDef by hand after closeSession (for some reason I was under the impression closeSession deletes the associated graphs). I will open a PR to fix this. I don't seem to observe any detectable effect on the memory, though.

Reply:
fix merged in https://github.com/cms-sw/cmssw/pull/32945